File: tobiasHe/assignments/twopointers/lc1/lc1.js
// Problem Statement #
// Given an array of sorted numbers and a target sum, find a pair in the array whose sum is equal to the given target.

// Write a function to return the indices of the two numbers (i.e. the pair) such that they add up to the given target.

function pair_with_target_sum(arr, targetSum) {
	// TODO:write code here
	// since it's a sorted array, two pointers can be used to keep track of where our two numbers are to sum up to our target
	// continue until left < right is not true. When it is true that means every single element has been met
	// sum of left and right = to targetSum, return left, right
	// if sum < targetSum then increment left pointer
	// else decrement right pointer

	// empty return []
	// nothing adds up to targetSum return []

	let leftPointer = 0;
	let rightPointer = arr.length - 1;
	let sum = 0;

	if (!arr.length) {
		return [];
	}

	while (leftPointer < rightPointer) {
		sum = arr[leftPointer] + arr[rightPointer];
		if (sum === targetSum) {
			return [ leftPointer, rightPointer ];
		} else if (sum < targetSum) {
			leftPointer++;
		} else {
			rightPointer--;
		}
	}
	return [];
}

console.log(pair_with_target_sum([ 1, 2, 3, 4, 6 ], 6));
console.log(pair_with_target_sum([ 2, 5, 9, 11 ], 11));
console.log(pair_with_target_sum([ 1, 2, 3, 4, 5 ], 221));
console.log(pair_with_target_sum([], 221));

// solution
// -----
// let left = 0,
// right = arr.length - 1;
// while (left < right) {
// const currentSum = arr[left] + arr[right];
// if (currentSum === targetSum) {
//   return [left, right];
// }

// if (targetSum > currentSum) {
//   left += 1; // we need a pair with a bigger sum
// } else {
//   right -= 1; // we need a pair with a smaller sum
// }
// }
// return [-1, -1];

File: roseWong/assignments/twopointers/lc1/lc1.py
# Problem Statement #
# Given an array of sorted numbers and a target sum,\
# find a pair in the array whose sum is equal to the given target.

# Write a function to return the indices of the two numbers (i.e. the pair)
# such that they add up to the given target.
# Psuedocode
""" 
if empty array, return empty array
initiate left pointer at index zero and right pointer at index array length - 1
sum left pointer and right pointer
if sum = target, return left ptr, right ptr in array format
if sum > target, decrement rt ptr
if sum < target, increment left ptr
do all if above while left < rt
"""

def pair_with_targetsum(arr, target_sum):
  leftPtr = 0
  rightPtr = len(arr)-1
  if (arr == []): # single = is assign, == is compare
      return []
  while leftPtr < rightPtr:
    sum = arr[leftPtr] + arr[rightPtr]
    if sum == target_sum:
      return [leftPtr, rightPtr]
    if sum > target_sum:
      rightPtr = rightPtr - 1
    else:
      leftPtr = leftPtr + 1
  return

def main():
  print(pair_with_targetsum([1, 2, 3, 4, 6], 6))
  """ 
1 + 6 = 7 > target => decrement rt = 4
1 + 4 = 5 < target +> increment left = 2
2 + 4 = 6 = target, return [1, 3]
  """
  print(pair_with_targetsum([2, 5, 9, 11], 11))

main()



# Solution
# -----
#   left, right = 0, len(arr) - 1
#   while(left < right):
#     current_sum = arr[left] + arr[right]
#     if current_sum == target_sum:
#       return [left, right]

#     if target_sum > current_sum:
#       left += 1  # we need a pair with a bigger sum
#     else:
#       right -= 1  # we need a pair with a smaller sum
#   return [-1, -1]
